### PR TITLE
Routing table rebuild optimizations

### PR DESCRIPTION
**Pull request by jfim**
Improve the performance of keeping the routing tables up to date for
large clusters with many tables by avoiding excessive external view
fetches. Fix instance configuration changes being ignored when
updating the routing tables.

### CODE REVIEW

**Comment by jfim** [comment]
Haven't yet tested the performance changes on a large cluster and this needs some additional tests.

**Comment by jfim** [comment]
There's a couple additional optimizations that could be done as well:
- Deferring updates for a couple of seconds so that we lower the overall number of ZK reads/stat calls
- Checking the instance config stats on external view change prior to fetching them

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 1]
```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+      LOGGER.info("Routing table for table {} requires rebuild due to empty/null previous instance configs", tableName);
+      return true;
+    }
+
+    // Gather relevant incoming instance configs
+    Map<String, InstanceConfig> currentRelevantInstanceConfigs = new HashMap<>();
+    for (InstanceConfig incomingInstanceConfig : instanceConfigs) {
+      String instanceName = incomingInstanceConfig.getInstanceName();
+
+      if (lastKnownInstanceConfigs.containsKey(instanceName)) {
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Should this be !containsKey()?

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 2]
```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+        currentRelevantInstanceConfigs.put(instanceName, incomingInstanceConfig);
       }
+    }
 
+    // Did some instances lose their configuration?
+    if (lastKnownInstanceConfigs.size() != currentRelevantInstanceConfigs.size()) {
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Is this only meant to be a quick check? It is possible that the number is the same but the instance config keys are different, right?

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 3]
```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+      if (instancesChanged) {
+        LOGGER.info(
+            "Routing table for table {} requires rebuild due to at least one instance changing state (instance {} enabled: {} -> {}; shutting down {} -> {})",
+            tableName, instanceName, wasEnabled, isEnabled, wasShuttingDown, isShuttingDown);
+        return true;
+      }
+    }
+
+    // No relevant changes, no need to update the routing table
+    LOGGER.info("Routing table for table {} does not require a rebuild");
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Worth having a metric on number of rebuilds per table?

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 4]
```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+
+      String wasShuttingDown =
+          previousInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+      String isShuttingDown =
+          currentInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+
+      boolean instancesChanged =
+          !EqualityUtils.isEqual(wasEnabled, isEnabled) || !EqualityUtils.isEqual(wasShuttingDown, isShuttingDown);
+
+      if (instancesChanged) {
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Am I right in that most of the time this will be 'false', because instanceConfig will change even if a segment is added to a server? If so, is there any way we can avoid notification for those changes? Not sure if there is a helix api for that.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 5]
```diff
@@ -229,13 +370,144 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
     } catch (Exception e) {
       LOGGER.error("Failed to update the TimeBoundaryService", e);
     }
+
+    LOGGER.info("Routing table update for table {} completed in {} ms", tableName,
+        (System.currentTimeMillis() - startTimeMillis));
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: worth adding a metric?

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 6]
```diff
@@ -229,13 +370,144 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+        String currentTableName = tablesServed.get(i);
+        int currentExternalViewVersion = externalViewStat.getVersion();
+        int lastKnownExternalViewVersion = _lastKnownExternalViewVersionMap.get(currentTableName);
+
+        if (lastKnownExternalViewVersion != currentExternalViewVersion) {
+          tablesThatChanged.add(currentTableName);
+        }
+      }
+    }
+
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: will be good to add a log message of how many tables we think need update ("N out of M")

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 7]
```diff
@@ -229,13 +370,144 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+      for (String tableThatChanged : tablesThatChanged) {
+        // We ignore the external views given by Helix on external view change and fetch the latest version as our
+        // version of Helix (0.6.5) does not batch external view change messages.
+        ExternalView externalView = helixDataAccessor.getProperty(propertyKeyBuilder.externalView(tableThatChanged));
+        if (isRoutingTableRebuildRequired(tableThatChanged, externalView, instanceConfigs)) {
+          buildRoutingTable(tableThatChanged, externalView, instanceConfigs);
+        }
+      }
+    }
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: And a log message here with how many tables we ended up updating (N out of M). Maybe good to have metrics also?

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 8]
```diff
@@ -229,13 +370,144 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+    if (!impactedTables.isEmpty()) {
+      List<InstanceConfig> instanceConfigs = helixDataAccessor.getChildValues(propertyKeyBuilder.instanceConfigs());
+
+      for (String tableName : impactedTables) {
+        ExternalView externalView = helixDataAccessor.getProperty(propertyKeyBuilder.externalView(tableName));
+        if (isRoutingTableRebuildRequired(tableName, externalView, instanceConfigs)) {
+          buildRoutingTable(tableName, externalView, instanceConfigs);
+        }
+      }
+    }
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: also good to have an N out of M routing table updates log here

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 9]
```diff
@@ -229,13 +370,144 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+        if (lastKnownExternalViewVersion != currentExternalViewVersion) {
+          tablesThatChanged.add(currentTableName);
+        }
+      }
+    }
+
+    // Fetch the instance configs and update the routing tables for the tables that changed
+    if (!tablesThatChanged.isEmpty()) {
+      List<InstanceConfig> instanceConfigs = helixDataAccessor.getChildValues(propertyKeyBuilder.instanceConfigs());
+
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: we could update the instance config versions of these instances so that we know we have processed them?

File: pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/BaseClusterIntegrationTest.java [hunk 10]
```diff
@@ -1237,12 +1237,12 @@ protected void ensurePinotConnectionIsCreated() {
...
-    if (resultSetGroup.getResultSetCount() > 0) {
-      return resultSetGroup.getResultSet(0).getInt(0);
-    }
-    return 0;
+      com.linkedin.pinot.client.ResultSetGroup resultSetGroup =
+          _pinotConnection.execute("SELECT COUNT(*) from " + tableName + " LIMIT 0");
+      if (resultSetGroup.getResultSetCount() > 0) {
+        return resultSetGroup.getResultSet(0).getInt(0);
+      }
+      return 0;
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/BaseClusterIntegrationTest.java]: Please revert this part.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 11]
```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+      return;
+    }
+
+    // Build the routing table
+    buildRoutingTable(tableName, externalView, instanceConfigList);
+  }
+
+  private boolean isRoutingTableRebuildRequired(String tableName, ExternalView externalView,
+      List<InstanceConfig> instanceConfigs) {
+    // In unit tests, always rebuild the routing table
```
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: It was needed when `markDataResourceOnline` checked if there was a rebuild needed, since not all unit tests have a valid HelixManager and/or initialize everything. Originally, everything went through `markDataResourceOnline`, including external view changes, and as such, it needed a rebuild check. Now that `markDataResourceOnline` is unconditional and only depends on the external view and instance configs passed, this workaround isn't required.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 12]
```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
     if (externalView == null) {
       return;
     }
+
+    // Check if there are relevant changes to the state or if we can skip this update
+    if (!isRoutingTableRebuildRequired(tableName, externalView, instanceConfigList)) {
```
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Now yes.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 1 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: No, this piece of code gathers the new instance configs for which we have a previous version and puts it in a map instanceName -> instanceConfig, so that those two can be compared.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 2 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Yes, we check this at a later step (see "Did some instance change state?")

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 4 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Instance config won't change on every segment add, it shouldn't change very often. Instance config refers to the contents of `CONFIGS/PARTICIPANT` for which the only updates would be for `isShuttingDown` and `HELIX_ENABLED`

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 3 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Yes, I'll add it.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 5 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: That specific metric was added in #799

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 6 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Yes, I'll probably remove listing the tables that don't need update as well, since it's a very long list that gets written on each external view change.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 7 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Yes.

File: pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/BaseClusterIntegrationTest.java (hunk 10 above)
Comment (jfim) [inline pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/BaseClusterIntegrationTest.java]: Done

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 13]
```diff
@@ -176,76 +196,406 @@ public void shutdown() {
   public void markDataResourceOnline(String tableName, ExternalView externalView,
       List<InstanceConfig> instanceConfigList) {
     if (externalView == null) {
+      // It is possible for us to get a request to serve a table for which there is no ideal state. In this case, just
```
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Yes.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 14]
```diff
@@ -176,76 +196,406 @@ public void shutdown() {
...
-    if (_routingTableLastKnownZkVersionMap.containsKey(tableName)) {
-      long lastKnownZkVersion = _routingTableLastKnownZkVersionMap.get(tableName);
-      if (externalViewRecordVersion == lastKnownZkVersion) {
-        LOGGER.info(
-            "No change on routing table version (current version {}, last known version {}), do nothing for table {}",
-            externalViewRecordVersion, lastKnownZkVersion, tableName);
-        return;
+    int lastKnownExternalViewVersion = _lastKnownExternalViewVersionMap.get(tableName);
+
+    if (externalViewRecordVersion != lastKnownExternalViewVersion || lastKnownExternalViewVersion == INVALID_EXTERNAL_VIEW_VERSION) {
```
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: In the unlikely event that the external view znode version is equal to `Integer.MIN_VALUE`, we'll always rebuild the routing table in this case. It shouldn't happen, but if it does, then we won't lose that routing table update.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 15]
```diff
@@ -176,76 +196,406 @@ public void shutdown() {
...
+      // Check if the instance got enabled/disabled or started/stopped shutting down since the last update
+      boolean wasEnabled = previousInstanceConfig.getInstanceEnabled();
+      boolean isEnabled = currentInstanceConfig.getInstanceEnabled();
+
+      String wasShuttingDown =
+          previousInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+      String isShuttingDown =
+          currentInstanceConfig.getRecord().getSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS);
+
+      boolean instancesChanged =
```
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: The code already checks for that specific condition seven lines above (see "If it's the same znode, don't bother comparing the contents of the instance configs").

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 16]
```diff
@@ -176,76 +196,406 @@ public void shutdown() {
...
+          !EqualityUtils.isEqual(wasEnabled, isEnabled) || !EqualityUtils.isEqual(wasShuttingDown, isShuttingDown);
+
+      if (instancesChanged) {
+        LOGGER.info(
+            "Routing table for table {} requires rebuild due to at least one instance changing state (instance {} enabled: {} -> {}; shutting down {} -> {})",
+            tableName, instanceName, wasEnabled, isEnabled, wasShuttingDown, isShuttingDown);
+        return true;
+      } else {
+        // Update the instance config in our last known instance config, since it hasn't changed
```
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: If the newly seen instance hasn't changed in a way that is considered different (eg. some field that we don't care about), then we keep it as the last seen version.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 17]
```diff
@@ -176,76 +196,406 @@ public void shutdown() {
...
+    Stat[] externalViewStats =
+        helixDataAccessor.getBaseDataAccessor().getStats(externalViewPaths, AccessOption.PERSISTENT);
+
+    long statsTime = System.currentTimeMillis();
+
+    // Make a list of external views that changed
+    List<String> tablesThatChanged = new ArrayList<>();
+
+    for (int i = 0; i < externalViewStats.length; i++) {
+      Stat externalViewStat = externalViewStats[i];
```
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: It would indicate in this case that the external view got deleted for some reason. `ZooKeeper#exists` can return null, see http://zookeeper.apache.org/doc/r3.4.6/api/org/apache/zookeeper/ZooKeeper.html#exists(java.lang.String, org.apache.zookeeper.Watcher)

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 15 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: Is it, because if we're at that point in the code, the znode versions are different. For reference, the code above is:

```
      // If it's the same znode, don't bother comparing the contents of the instance configs
      if (previousInstanceConfig.getRecord().getVersion() == currentInstanceConfig.getRecord().getVersion()) {
        continue;
      }
```

Hence we can only be at that place in the code if the znodes are different.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java (hunk 1 above)
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: In that case, if a new instanceconfig gets added, we are not doing anything about it?

Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: We don't care because our routing table does not contain this new instance. If a new instance is added to serve this table, we will get an external view change, at which point we'll keep track of this new instance's configuration.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java [hunk 18]
```diff
@@ -176,49 +195,171 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
...
+
+    if (externalViewRecordVersion != lastKnownExternalViewVersion) {
+      LOGGER.info(
+          "Routing table for table {} requires rebuild due to external view change (current version {}, last known version {})",
+          tableName, externalViewRecordVersion, lastKnownExternalViewVersion);
+      return true;
+    }
+
+    // Check if there are relevant instance config changes
+    Map<String, InstanceConfig> lastKnownInstanceConfigs = _lastKnownInstanceConfigsForTable.get(tableName);
```
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java]: It just keeps track of the previous instance configs and doesn't rebuild if there are no instance config changes or if the instance config changes are unrelated to `isShutdownInProgress` and `HELIX_ENABLED`. There's unfortunately not a very good compact way to express this in Java.

**Comment by jfim** [comment]
Some unit tests failed as they used `Thread.sleep()` and expected the cluster to reach a given state. The fix for them is in #834.
